include/fusilli/node/matmul_node.h

```diff
@@ -0,0 +1,222 @@
+    int64_t aK = aDim[aRank - 1]; // Last dimension of A
+    int64_t bK = bDim[bRank - 2]; // Second-to-last dimension of B
+
+    FUSILLI_RETURN_ERROR_IF(
+        aK != bK, ErrorCode::InvalidAttribute,
+        "Matmul input tensors A and B have incompatible inner dimensions (K): "
+        "A has K=" +
+            std::to_string(aK) + ", B has K=" + std::to_string(bK));
+
+    // Check that batch dimensions are broadcastable.
+    // Since both inputs have the same rank, we can directly compare batch dims.
+    size_t batchDims = aRank - kNonBatchRank;
+    for (size_t i = 0; i < batchDims; ++i) {
+      int64_t aDimVal = aDim[i];
+      int64_t bDimVal = bDim[i];
+      FUSILLI_RETURN_ERROR_IF(
+          !(aDimVal % bDimVal == 0 || bDimVal % aDimVal == 0),
+          ErrorCode::InvalidAttribute,
+          "Matmul input tensors A and B have incompatible batch dimensions for "
+          "broadcasting at index " +
+              std::to_string(i) + ": A has dim=" + std::to_string(aDimVal) +
+              ", B has dim=" + std::to_string(bDimVal));
+    }
+
+    // Check that batch dimensions are outermost and non-transposed.
+    // This is equivalent to checking that perm[i] == i for all batch dims.
+    auto checkBatchDims = [&](const std::shared_ptr<TensorAttr> &tensor,
+                              const std::string &name) {
+      std::vector<int64_t> perm = tensor->getLogicalToPhysicalPermuteOrder();
+      for (size_t i = 0; i < batchDims; ++i) {
+        FUSILLI_RETURN_ERROR_IF(
+            perm[i] != static_cast<int64_t>(i), ErrorCode::InvalidAttribute,
+            "Matmul input tensor " + name +
+                " has batch dimensions that are not outermost or are "
+                "transposed");
+      }
+      return ok();
+    };
+
+    FUSILLI_CHECK_ERROR(checkBatchDims(aT, "A"));
+    FUSILLI_CHECK_ERROR(checkBatchDims(bT, "B"));
+
+    return ok();
+  }
+
+  ErrorObject inferPropertiesNode() override final {
+    FUSILLI_LOG_LABEL_ENDL("INFO: Inferring properties for MatmulNode '"
+                           << matmulAttr.getName() << "'");
+
+    matmulAttr.fillFromContext(context);
+
+    std::shared_ptr<TensorAttr> aT = matmulAttr.getA();
+    std::shared_ptr<TensorAttr> bT = matmulAttr.getB();
+    std::shared_ptr<TensorAttr> cT = matmulAttr.getC();
+
+    const std::vector<int64_t> &aDim = aT->getDim();
+    const std::vector<int64_t> &bDim = bT->getDim();
+
+    const std::vector<int64_t> &cDim = cT->getDim();
+    const std::vector<int64_t> &cStride = cT->getStride();
+
+    // Infer shape of output tensor.
+    if (cDim.empty())
+      cT->setDim(getMatmulInferredOutputShape(aDim, bDim));
+
+    // Output stride is contiguous (row-major) when unspecified.
+    if (cStride.empty()) {
+      cT->setStride(
+          generateStrideFromDim(cDim, getContiguousStrideOrder(cDim.size())));
+    }
+
+    return ok();
+  }
+
+  ErrorObject postValidateNode() const override final {
```

Comment:
Can you replicate the `checkBatchDims` test for `cT` as well here? This will be useful for cases where users annotate the stride incorrectly and use transposes on batch dims.